File: Go/Day25.go
package main

import AH "./adventhelper"

// b^a mod p
func modPow(b int, a int, p int) (x int) {
	x = 1;

	for ; a > 0; {
		if (a % 2 == 1) {
			a -= 1
			x *= b
			x %= p
		}
		a /= 2
		b *= b
		b %= p
	}

	return
}

// b^a = x mod p
func dicreteLog(b int, x int, p int) (a int) {
	temp := b
	for a = 1; ; a++ {
		if temp == x {
			break
		}
		temp *= b
		temp %= p
	}
	return a
}

func main() {
	is, _ := AH.ReadIntFile("../input/input25.txt")

	AH.PrintSoln(25, modPow(is[0], dicreteLog(7, is[1], 20201227), 20201227),
	                 "See you next year, I love you.")

	return
}

File: Haskell/Day25.hs
import AdventHelper

-- b^a mod p
modPow :: Integer -> Integer -> Integer -> Integer
modPow b 0 p = 1
modPow b 1 p = mod b p
modPow b a p | even a = mod ((modPow b (div a 2) p) ^ 2) p
             | odd  a = mod ((modPow b (div (a-1) 2) p) ^ 2 * b) p

-- a*b mod p
modMult :: Integer -> Integer -> Integer -> Integer
modMult p a b = mod (a * b) p

-- b^a = x mod p <-- find a
discreteLog :: Integer -> Integer -> Integer -> Integer
discreteLog b x p = 1 + fromIntegral (length $ takeWhile (/= x) $ scanl1 (modMult p) (repeat b))

main :: IO ()
main = do
  putStrLn "Day 25"
  f <- readFile "../input/input25.txt"
  let [cardPub, doorPub] = map read (lines f) :: [Integer]

  let doorLoop = discreteLog 7 doorPub 20201227

  printSoln 1 $ modPow cardPub doorLoop 20201227
  printSoln 2 "See you next year, I love you."
